### ISSUE TITLE
Handle multiple subnets per building

### ISSUE THREAD

**Issue by lukasjuhrich** [agdsn/pycroft#82]
Currently, there are buildings which are assigned multiple subnets.  This means that either the model has to be changed in order to support more than just the default_subnet, or the user creation workflow has to enable manual selection of a subnet to use.

**Comment by lukasjuhrich** [comment]
The current Idea would be to create a relation `subnet_availability` which realizes an n:n mapping from `switch_interface` to `subnet`.

One might ask why subnets are coupled to SwitchInterfaces and not switches, as the latter option would drastically reduce the tuples in the `subnet_availability` relation. However, this design decision has been manifested previously in the `SwitchInterface.default_subnets` property -- @sebschrader, do you know what was the rationale to assign subnets to `SwitchInterface`s rather than `Switch`es?

**Comment by lukasjuhrich** [comment]
Fixed by 186b1bb6b705cc73698a8e2ca637d396327f8b51, 8fea28eccac873c31205827254e479d72a7caecb and others in that range.

**Comment by sebschrader** [comment]
Sorry for getting back to you so late. I had your request in my queue, but never came around to giving you an answer.

A `SwitchInterface` is not a switch port, as you probably think (deducing from your example in 8fea28eccac873c31205827254e479d72a7caecb), it is rather the logical network interface of the switch, that has a MAC address and may have a IP addresses. Think `Vlan-Interface` on ComWare in `bridge` mode (but may also be an physical interface if in `route` mode) devices or simply the `vlan` on ProVision/ArubaOS devices.

Therefore there shouldn't be an explosion of tuples, most switches will need only one `SwitchInterface` modeled in Pycroft currently.

**Comment by lukasjuhrich** [comment]
@sebschrader This contradicts with the data the current importer sets up and how the ports are [displayed](https://github.com/agdsn/pycroft/blob/6cb59a9e0429ba02cdb981f942e8da50ad592581/web/blueprints/infrastructure/__init__.py#L79): Every switch has a switch interface for each physical port (plus e.g. a management interface). Consequently, the relevant subnets have to be provided very often and in a redundant manner.

Your description seems to imply that instead of a 1:1-mapping (in most cases) from a `SwitchPatchPort` to a `SwitchInterface`, we should have only one or two `SwitchInterfaces` which are themselves connected to a ton of `SwitchPatchPort`s (for each room connected to the switch). Is that correct?

With your proposition, the following questions would arise:

- Do we need one `SwitchInterface` per subnet being available to a switch? If so, how is decided which ports are patched to which `SwitchPatchPort`s? (The latter would imply that per room, only one subnet can be chosen, which is bad)
- How would we access the Hades Logs for a room? We *need* to keep track of (the name of) the physical port a room is patched to.

**Comment by sebschrader** [comment]
We already talked about this via different means. I just want to leave this comment here for documentation.

Your current approach is sufficient, but if you would want to model it correctly, we would need a `SwitchPort` class and have the following relations:

```
SwitchPatchPort 1 <-> 1 SwitchPort n <-> 1 SwitchInterface n <-> m Subnet
                                                           n <-> 1 VLAN
                                                           1 <-> 1 IP
```

The current approach is sufficient for now however.

**Comment by lukasjuhrich** [comment]
I think we should keep this in mind. @sebschrader, what do you think - reopen or open a new issue?

**Comment by sebschrader** [comment]
Well, this issue touches multiple things, so I would suggest reopening it and maybe creating another issue.

It should be reopened because subnet selection needs to be reworked. It must be based on your port and not your building. There are already multiple subnets per building in some cases and going forward with our future L3 architecture there will be a subnet per port anyway.

One might track the introduction of a new `SwitchPort` in a new issue, but I'm not sure if we want to this right now.

Furthermore I realized that the `SwitchPort` and `SwitchInterface` might actually belong to different devices. Users are connected to an access switch but their subnet is routed by a different device. This is the case both with our current L2 architecture and our future L3 architecture. One might argue that this level of detail is not needed and we can get away with some simplifications. We could for example drop the `SwitchInterface` altogether and replace it with the aforementioned `SwitchPort` object and associate the `SwitchPort` with a subnet. We would lose the capability to store the MAC addresses and IP addresses of our switches and routers in Pycroft, but we might not need that. We don't need the MAC addresses of these devices in Pycroft anyway. We should probably store the IP addresses of our devices in Pycroft, so that we can list them as allocated.

**Comment by lukasjuhrich** [comment]
> We would lose the capability to store the MAC addresses and IP addresses of our switches and routers in Pycroft, but we might not need that. We don't need the MAC addresses of these devices in Pycroft anyway. We should probably store the IP addresses of our devices in Pycroft, so that we can list them as allocated.

Just an addition, for interoperability with Hades, the IP address of the access must be preserved. But I agree that we should rework this to e.g. Remove the MAC-addresses.

However, at the moment I don't entirely understand what the implications of a NAT infrastructure would be. That remains to be discussed.

**Comment by sebschrader** [comment]
> Just an addition, for interoperability with Hades, the IP address of the access must be preserved. But I agree that we should rework this to e.g. Remove the MAC-addresses.

For Hades interoperability the NAS-IP-Address of the access switch must be known. The NAS-IP-Address is the management IP address of the switch and not the IP address of the router in the users subnet.

**Comment by lukasjuhrich** [comment]
> The NAS-IP-Address is the management IP address of the switch and not the IP address of the router in the users subnet

Ah, you're right, I meant the management-IP.

Note however that because the data we have on our legacy system is not complete, the switches' IP addresses are not imported:

```
In [19]: [(s.name, s.ips) for s in switches]                                                                           
Out[19]:                                                                                                               
[('switch-wu1-06', []),                                    
 ('switch-zw41', []),                                                                                                  
 ('switch-zw41a', []),                                                                                                                                                                                                                         
 ('switch-zw41b', []),                                     
 ('switch-wu11-04', []),                                                                                               
 ('switch-wu11-12', []),                                   
 ('switch-wu9-04', []),                                                                                                
 ('switch-wu9-11', []),                                                                                                                                                                                                                        
 ('switch-wu7-07', []),                                                                                                
 ('switch-wu3-04', []),                                                                                                
 ('switch-wu3-11', []),                                    
 ('switch-wu1-03', []),                                                                                                
 ('switch-wu7-03', []),                                                                                                                                                                                                                        
 ('switch-zw41c', []),                                                                                                                                                                                                                         
 ('switch-zw41d', []),                                     
 ('switch-wu5-00', []),                                    
 ('switch-wu5-07', []),                                                                                                
 ('switch-wu5-13', []),                                    
 ('switch-wu7-13', []),                                                                                                
 ('switch-wu1-12', []),
 ('switch-bor-04', []),                                                                                                                                                                                                                        
 ('switch-bor-03', []),                                    
 ('switch-bor-05', []),                                    
 ('switch-bor-06', []),                                                                                                
 ('switch-bor-07', []),                                                                                                
 ('switch-bor-02', []),                                                                                                
 ('switch-zeu-saniert', []),                               
 ('switch-zeu-unsaniert', []),                             
 ('switch-bu22', []),                                                                                                  
 ('switch-bu24', []),                                      
 ('switch-bor-01', [])]
```
The only fact that a `Switch` _may_ have IP addresses is because in the current Model, it inherits from `Host`.

**Comment by sebschrader** [comment]
Obtaining the management IP of the switches this way is not reliable. What if the `Switch` has multiple IP addresses in different `Subnet`s? Management IP is also a rather broad thing. We might use different IP addresses for SSH access and RADIUS.

It would probably be the best, if we're adding a `radius_ip` or `nas_ip_address` relationship to `Switch`, which links it to an `IP`. This relationship would be 1:1.

**Comment by sebschrader** [comment]
Obtaining the management IP of the switches this way is not reliable. What if the `Switch` has multiple IP addresses in different `Subnet`s? We do have a `management_ip` property, which you didn't use in your example. Management IP is however a rather broad thing. We might use different IP addresses for SSH access and RADIUS. Here for example are all the different services on our ProCurve switches for which we could specify different source-interfaces:

```
# ip source-interface ?
 radius                The RADIUS protocol.
 sntp                  The SNTP protocol.
 syslog                The syslog protocol.
 tacacs                The TACACS+ protocol.
 telnet                The Telnet protocol.
 tftp                  The TFTP protocol.
 sflow                 The sFlow protocol.
 all                   All protocols above.
```

It would probably be the best, if we're adding a `radius_ip` or `nas_ip_address` relationship to `Switch`, which links it to an `IP`. This relationship would be 1:1.

**Comment by lukasjuhrich** [comment]
Everything we talked about here has been fixed by the changes around #147 and various simplifications that have been made (removal of some polymorphisms).
